pegasus/helpers/hoc_helpers.rb

```diff
@@ -2,23 +2,67 @@
 require 'cdo/graphics/certificate_image'
 require 'dynamic_config/gatekeeper'
 
-def hoc_activity_writes_disabled
-  Gatekeeper.allows('hoc_activity_writes_disabled')
+UNSAMPLED_SESSION_ID = 'HOC_UNSAMPLED'
+
+# Creates a session row and sets the hour of code cookie to the session_id,
+# if the user is assigned to the sample set (as decided by a random choice
+# based on the hoc_activity_sample_proportion DCDO variable).
+#
+# If, however, the user is not in the sample, returns nil and sets the cookie
+# to UNSAMPLED_SESSION_ID.
+#
+# The "weight" encoded in the session row is set to 1/p, where p is the
+# proportion of sessions in the sample, so that reports can compute the
+# approximate number of actual sessions by summing over the weights.
+def create_session_row_unless_unsampled(row)
+  # We don't need to do anything if we've already decided this session is unsampled.
+  return if unsampled_session?
+
+  # Decide whether the session should be sampled.
+  p = DCDO.get('hoc_activity_sample_proportion', 1.0).to_f
+  if Kernel.rand < p  # If p=0, no sessions are in the sample.
+    # If we decided to make the session sampled, create the session row and set the hoc cookie.
+    row = create_session_row(row, weight: 1.0 / p)
+  else
+    # Otherwise set the hoc cookie to make the session as unsampled.
+    set_hour_of_code_cookie_for_row(session: UNSAMPLED_SESSION_ID)
+    row = nil
+  end
+  row
 end
 
-def create_session_row(row)
+# Creates a session row with the given weight and sets the hour of code cookie to contain
+# the session id.
+def create_session_row(row, weight: 1.0)
   retries = 3
 
   begin
-    row[:session] = SecureRandom.hex
+    row[:session] = create_session_id(weight)
     row[:id] = DB[:hoc_activity].insert(row)
   end while row[:id] == 0 && (retries -= 1) > 0
 
   raise "Couldn't create a unique session row." if row[:id] == 0
-
+  set_hour_of_code_cookie_for_row(row)
   row
 end
 
+# Create a session id that also encodes the weight of the session.
+# We should actually use a separate column for the weight, but need to defer adding
+# that column until after the hour of code. (hoc_activity currently has ~100M rows).
+def create_session_id(weight)
+  "_#{weight}_#{SecureRandom.hex}"
```

Comment:
This is a good point!  The session column is a VARCHAR(50) and the hex string is 32 characters.  That gives 16 characters for the weight after allowing room for the underscores.  We could just exceed that in pathological cases.  I added a round(6) call so that the weight doesn't take too many characters.